pydatastructs/strings/suffix_tree.py

```diff
@@ -0,0 +1,273 @@
+        else:
+            stringIdxs = set(stringIdxs)
+        deepestNode = self._find_lcs(self.root, stringIdxs)
+        start = deepestNode.idx
+        end = deepestNode.idx + deepestNode.depth
+        return self.word[start:end]
+
+    def _find_lcs(self, node, stringIdxs):
+        nodes = [self._find_lcs(n, stringIdxs)
+                 for n in node.transition_links.values()
+                 if n.generalized_idxs.issuperset(stringIdxs)]
+        if nodes == []:
+            return node
+        deepestNode = max(nodes, key=lambda n: n.depth)
+        return deepestNode
+
+    def _generalized_word_starts(self, xs):
+        self.word_starts = []
+        i = 0
+        for n in range(len(xs)):
+            self.word_starts.append(i)
+            i += len(xs[n]) + 1
+
+    def find(self, y):
+        """
+        Finds the starting position of the substring y in the string used for
+        building the Suffix tree.
+
+        Parameters
+        ==========
+
+        y: str
+
+        Returns
+        =======
+
+        Index of the starting position of string y in the string used for building the Suffix tree
+        -1 if y is not a substring.
+        """
+        node = self.root
+        while True:
+            edge = self._edgeLabel(node, node.parent)
+            if edge.startswith(y):
+                return node.idx
+
+            i = 0
+            while (i < len(edge) and edge[i] == y[0]):
+                y = y[1:]
+                i += 1
+
+            if i != 0:
+                if i == len(edge) and y != '':
+                    pass
+                else:
+                    return -1
+
+            node = node._get_transition_link(y[0])
+            if not node:
+                return -1
+
+    def find_all(self, y):
+        """
+        Finds the starting position of the substring y in the string used for
+        building the Suffix tree.
+
+        Parameters
+        ==========
+
+        y: str
+
+        Returns
+        =======
+
+        Set of Index of the starting positions of string y in the string used for building the Suffix tree
+        {} if y is not a substring.
+        """
+        node = self.root
+        while True:
+            edge = self._edgeLabel(node, node.parent)
+            if edge.startswith(y):
+                break
+            i = 0
+            while (i < len(edge) and edge[i] == y[0]):
+                y = y[1:]
+                i += 1
+            if i != 0:
+                if i == len(edge) and y != '':
+                    pass
+                else:
+                    return {}
+            node = node._get_transition_link(y[0])
+            if not node:
+                return {}
+
+        leaves = node._get_leaves()
+        return {n.idx for n in leaves}
+
+    def _edgeLabel(self, node, parent):
+        return self.word[node.idx + parent.depth: node.idx + node.depth]
+
+    def _terminalSymbolsGenerator(self):
```

Comment:
What is the logic of this method? How it is doing it's job?

Reply:
in range from Hexadecimal (57344 - 1114109) where it can accommodate the nodes of these range and generate symbols for the suffix tree with ascii as terminal symbols. The yield helps optimizing the code that it don't have to initialize the UPPA value again

Reply:
These ranges are used just to make sure that the symbols aren't part of the suffix tree inputs